routing/missioncontrol_store.go

```diff
@@ -0,0 +1,235 @@
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+var (
+	// resultsKey is the fixed key under which the attempt results are
+	// stored.
+	resultsKey = []byte("missioncontrol-results")
+
+	// Big endian is the preferred byte order, due to cursor scans over
+	// integer keys iterating in order.
+	byteOrder = binary.BigEndian
+
+	// unknownFailureSourceIdx is the database encoding of an unknown error
+	// source.
+	unknownFailureSourceIdx = -1
+)
+
+// missionControlStore is a bolt db based implementation of a mission control
+// store. It stores the raw payment attempt data from which the internal mission
+// controls state can be rederived on startup. This allows the mission control
+// internal data structure to be changed without requiring a database migration.
+// Also changes to mission control parameters can be applied to historical data.
+// Finally, it enables importing raw data from an external source.
+type missionControlStore struct {
+	db         *bbolt.DB
+	maxRecords int
+}
+
+func newMissionControlStore(db *bbolt.DB, maxRecords int) (*missionControlStore, error) {
+	// Create buckets if not yet existing.
+	err := db.Update(func(tx *bbolt.Tx) error {
+		_, err := tx.CreateBucketIfNotExists(resultsKey)
+		if err != nil {
+			return fmt.Errorf("cannot create results bucket: %v",
+				err)
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return &missionControlStore{
+		db:         db,
+		maxRecords: maxRecords,
+	}, nil
+}
+
+// clear removes all results from the db.
+func (b *missionControlStore) clear() error {
+	return b.db.Update(func(tx *bbolt.Tx) error {
+		if err := tx.DeleteBucket(resultsKey); err != nil {
+			return err
+		}
+
+		_, err := tx.CreateBucket(resultsKey)
+		return err
+	})
+}
+
+// fetchAll returns all results currently stored in the database.
+func (b *missionControlStore) fetchAll() ([]*paymentResult, error) {
+	var results []*paymentResult
+
+	err := b.db.View(func(tx *bbolt.Tx) error {
+		resultBucket := tx.Bucket(resultsKey)
+		results = make([]*paymentResult, resultBucket.Stats().KeyN)
+
+		i := 0
+		return resultBucket.ForEach(func(k, v []byte) error {
+			result, err := deserializeResult(k, v)
+			if err != nil {
+				return err
+			}
+
+			results[i] = result
+			i++
+
+			return nil
+		})
+
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	return results, nil
+}
+
+// deserializeResult deserializes a payment result.
+func deserializeResult(k, v []byte) (
+	*paymentResult, error) {
+
+	// Parse payment id.
+	result := paymentResult{
+		id: byteOrder.Uint64(k[8:]),
+	}
+
+	r := bytes.NewReader(v)
+
+	// Read timestamps, success status and failure source index.
+	var (
+		timeFwd, timeReply uint64
+		dbFailureSourceIdx int32
+	)
+
+	err := channeldb.ReadElements(
+		r, &timeFwd, &timeReply, &result.success, &dbFailureSourceIdx,
+	)
+	if err != nil {
+		return nil, err
+	}
+	result.timeFwd = time.Unix(0, int64(timeFwd)).UTC()
+	result.timeReply = time.Unix(0, int64(timeReply)).UTC()
```

Comment:
do we use this `UTC()` conversion anywhere else in the codebase?

Reply:
I did a quick find and it seems we don't. The reason I do this is to get UTC timezones for log messages that contain these times.

Reply:
UTC is is the default https://play.golang.org/p/e_-nuERpM66

Reply:
I think that's just because the server being used for the playground is on UTC. Running this locally results in my timezone being shown instead.

Reply:
hmm, that's obnoxious.

but still the bigger question is why wouldn't we print this in the user's local timezone? nothing else in our logs are UTCified

Reply:
I assumed we want UTC in order to be able to easily compare/merge logs from different machines without needing to think about physical location. But if we don't do it anywhere else in the code base, we may not want to do it here either.

Reply:
UTC removed